### PR TITLE
JPN-436 Insights module for the Community Page

### PR DESCRIPTION
**Pull request by d34th4ck3r**
### Links
- [JPN-436](https://wikia-inc.atlassian.net/browse/JPN-436)
### Description

This PR introduces the first of Insights based modules to the Special:CommunityPage and creates foundation for other modules to be added easily.
### Reviewers

@Wikia/spitfires @perjg

### CODE REVIEW

File: extensions/wikia/CommunityPage/CommunityPageSpecialController.class.php [hunk 1]
```diff
@@ -43,10 +48,71 @@ public function index() {
...
+					['href' => $insight['metadata']['lastRevision']['userpage']],
+					$insight['metadata']['lastRevision']['username']
+				),
+				$this->getLang()->userDate( $timestamp, $this->getUser() )
+			)->escaped();
+		}
+		return $insightsPages;
+	}
+
+	public function header() {
```
Comment (Grunny) [inline extensions/wikia/CommunityPage/CommunityPageSpecialController.class.php]: This seems unrelated, where is this from?

File: extensions/wikia/InsightsV2/services/InsightsService.class.php [hunk 2]
```diff
@@ -0,0 +1,40 @@
+<?php
+
+
+class InsightsService extends WikiaService
+{
```
Comment (Grunny) [inline extensions/wikia/InsightsV2/services/InsightsService.class.php]: Brace should be on the same line as the class definition per conventions.

File: extensions/wikia/InsightsV2/services/InsightsService.class.php [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+	/**
+	 * @param string $type type of model
+	 * @param int $size number of pages we need
+	 * @return array
+	 */
+	public function getInsightPages( $type, $size, $sortingType ) {
+		if ( InsightsHelper::isInsightPage( $type ) ) {
+			$model = InsightsHelper::getInsightModel( $type );
+			$insightData = ( new InsightsContext( $model ) )->fetchData();
+			$sortedInsightArticleIds = ( new InsightsSorting( $model->getConfig() ) )->getSortedData( $insightData, ['sort' => $sortingType] );
```
Comment (Grunny) [inline extensions/wikia/InsightsV2/services/InsightsService.class.php]: Line is too long.

File: extensions/wikia/InsightsV2/services/InsightsService.class.php [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+			$aritclesIds = $this->truncateTo( $sortedInsightArticleIds, $size );
+			return $this->getArticlesData( $insightData, $aritclesIds );
+		}
+		return [];
```
Comment (Grunny) [inline extensions/wikia/InsightsV2/services/InsightsService.class.php]: I'd consider reversing this so the early return of empty data is in an if block, and the main work of the method is unindented.

File: extensions/wikia/InsightsV2/services/InsightsService.class.php [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+	public function getArticlesData( $articles, $ids ) {
+		$content = [];
+		foreach ( $ids as $id ) {
+			$content[] = $articles[$id];
+		}
+		return $content;
+	}
+
+	public function truncateTo( $array, $size ) {
```
Comment (Grunny) [inline extensions/wikia/InsightsV2/services/InsightsService.class.php]: Should be a private method.

File: extensions/wikia/CommunityPage/CommunityPageSpecialController.class.php [hunk 6]
```diff
@@ -43,10 +48,71 @@ public function index() {
...
 			'recentActivityModule' => $this->getRecentActivityData(),
+			'popularPages' => $this->getInsightModule( 'popularpages' )
 		] );
 	}
 
 	/**
+	 * @param string $type type of module we want to build.
+	 * @return array Insight Module
+	 */
+
```
Comment (Grunny) [inline extensions/wikia/CommunityPage/CommunityPageSpecialController.class.php]: Remove this blank line.

File: extensions/wikia/CommunityPage/CommunityPageSpecialController.class.php [hunk 7]
```diff
@@ -43,10 +48,71 @@ public function index() {
...
+		return $this->addingLastRevision( $insightPages );
+	}
+
+	/**
+	 * @param array $insightsPages
+	 * @return array Prepare message about who and when last edited given article
+	 * @throws MWException
+	 */
+
```
Comment (Grunny) [inline extensions/wikia/CommunityPage/CommunityPageSpecialController.class.php]: And this one. :)

File: extensions/wikia/CommunityPage/CommunityPageSpecialController.class.php [hunk 8]
```diff
@@ -43,10 +48,71 @@ public function index() {
...
+	private function addingLastRevision( $insightsPages ) {
```
Comment (Grunny) [inline extensions/wikia/CommunityPage/CommunityPageSpecialController.class.php]: This would be better named as `addLastRevision`.

File: extensions/wikia/CommunityPage/CommunityPage.i18n.php [hunk 9]
```diff
@@ -103,6 +106,9 @@
 	'communitypage-no-admins-contact' => 'Contact us to become the first admin text',
 	'communitypage-communitycentral-link' => 'Link to the community page for this language',
 	'communitypage-no-members' => 'Text for zero state of all members list',
+	'communitypage-popularpages-title' => 'Title for the Top Rising pages module',
+	'communitypage-popularpages-description' => 'Description for the Top Rising pages module',
+	'communitypage-lastrevision' => 'Details about user that last edited and time of editing',
```
Comment (Grunny) [inline extensions/wikia/CommunityPage/CommunityPage.i18n.php]: Messages with variables should have the variables explicitly documented in the QQQ strings so translators know exactly what `$1` and `$2` are.

File: extensions/wikia/CommunityPage/CommunityPageSpecialController.class.php [hunk 10]
```diff
@@ -43,6 +48,66 @@ public function index() {
...
+			$insightsPages['pages'][$key]['lastRevision'] = $this->msg( 'communitypage-lastrevision' )->rawParams(
+				Html::element(
+					'a',
+					['href' => $insight['metadata']['lastRevision']['userpage']],
+					$insight['metadata']['lastRevision']['username']
+				),
+				$this->getLang()->userDate( $timestamp, $this->getUser() )
+			)->escaped();
+			$noOfView = $insight['metadata']['pv7'];
+			$insightsPages['pages'][$key]['pv7'] = $this->msg('communitypage-noofviews',$noOfView)->text();
```
Comment (Grunny) [inline extensions/wikia/CommunityPage/CommunityPageSpecialController.class.php]: The `msg` call should be spacey per conventions.

File: extensions/wikia/InsightsV2/services/InsightsService.class.php [hunk 11]
```diff
@@ -1,31 +1,34 @@
...
+		if ( !InsightsHelper::isInsightPage( $type ) ) {
+			return [];
+		}else{
 			$model = InsightsHelper::getInsightModel( $type );
 			$insightData = ( new InsightsContext( $model ) )->fetchData();
-			$sortedInsightArticleIds = ( new InsightsSorting( $model->getConfig() ) )->getSortedData( $insightData, ['sort' => $sortingType] );
+			$sortedInsightArticleIds = (
+			new InsightsSorting( $model->getConfig() ) )
+				->getSortedData( $insightData,
+					['sort' => $sortingType]
```
Comment (Grunny) [inline extensions/wikia/InsightsV2/services/InsightsService.class.php]: Should be spacey per conventions.

File: extensions/wikia/InsightsV2/services/InsightsService.class.php [hunk 12]
```diff
@@ -1,31 +1,34 @@
...
 	/**
 	 * @param string $type type of model
 	 * @param int $size number of pages we need
 	 * @return array
 	 */
 	public function getInsightPages( $type, $size, $sortingType ) {
-		if ( InsightsHelper::isInsightPage( $type ) ) {
+		if ( !InsightsHelper::isInsightPage( $type ) ) {
+			return [];
+		}else{
```
Comment (Grunny) [inline extensions/wikia/InsightsV2/services/InsightsService.class.php]: You have a return in the `if` block so it would be cleaner to remove the `else` and just unindent the main body of the method..

File: extensions/wikia/InsightsV2/services/InsightsService.class.php [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+	public function getInsightPages( $type, $size, $sortingType ) {
+		if ( !InsightsHelper::isInsightPage( $type ) ) {
+			return [];
+		}else{
+			$model = InsightsHelper::getInsightModel( $type );
+			$insightData = ( new InsightsContext( $model ) )->fetchData();
+			$sortedInsightArticleIds = (
+			new InsightsSorting( $model->getConfig() ) )
+				->getSortedData( $insightData,
+					['sort' => $sortingType]
```
Comment (Grunny) [inline extensions/wikia/InsightsV2/services/InsightsService.class.php]: Should be spacey per conventions.

File: extensions/wikia/InsightsV2/services/InsightsService.class.php [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+
+	/**
+	 * @param string $type type of model
+	 * @param int $size number of pages we need
+	 * @return array
+	 */
+	public function getInsightPages( $type, $size, $sortingType ) {
+		if ( !InsightsHelper::isInsightPage( $type ) ) {
+			return [];
+		}else{
```
Comment (Grunny) [inline extensions/wikia/InsightsV2/services/InsightsService.class.php]: You have a return in the `if` block so it would be cleaner to remove the `else` and just unindent the main body of the method.

File: extensions/wikia/InsightsV2/services/InsightsService.class.php [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+			$model = InsightsHelper::getInsightModel( $type );
+			$insightData = ( new InsightsContext( $model ) )->fetchData();
+			$sortedInsightArticleIds = (
+			new InsightsSorting( $model->getConfig() ) )
```
Comment (Grunny) [inline extensions/wikia/InsightsV2/services/InsightsService.class.php]: This shouldn't be on a new line from the opening `(`.

File: extensions/wikia/InsightsV2/services/InsightsService.class.php [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+			return $this->getArticlesData( $insightData, $aritclesIds );
+		}
+	}
+
+	/**
+	 * @param array $articles all articles data
+	 * @param array $ids list of article ids we need data for
+	 * @return array
+	 */
+	public function getArticlesData( $articles, $ids ) {
```
Comment (Grunny) [inline extensions/wikia/InsightsV2/services/InsightsService.class.php]: This can be private as well.

File: extensions/wikia/CommunityPage/CommunityPageSpecialController.class.php [hunk 17]
```diff
@@ -43,10 +48,56 @@ public function index() {
...
+					['href' => $insight['metadata']['lastRevision']['userpage']],
+					$insight['metadata']['lastRevision']['username']
+				),
+				$this->getLang()->userDate( $timestamp, $this->getUser() )
+			)->escaped();
+			$insightsPages['pages'][$key]['pageviews'] = $this->msg(
+				'communitypage-noofviews',
+				$insight['metadata']['pv7']
+			)->text();
+			$insightsPages['pages'][$key]['editlink'] = $insight['link']['url'] . '?action=edit';
```
Comment (Grunny) [inline extensions/wikia/CommunityPage/CommunityPageSpecialController.class.php]: We probably want to make sure we support VE for the edit link as well.

File: extensions/wikia/CommunityPage/templates/insightsPopularPages.mustache [hunk 18]
```diff
@@ -0,0 +1,14 @@
+<section class="insights-module">
+	<h3>{{popularPages.title}}</h3>
+	<p>{{popularPages.description}}</p>
+  <a href="{{popularPages.fulllistlink}}">{{popularPages.fulllist}}</a>
+	<ul>
+		{{#popularPages.pages}}
+			<li>
+				<a href="{{link.url}}">{{link.text}}</a><br>
+				<span>{{{lastRevision}}} • {{pageviews}}</span>
+        <a href="{{editlink}}">{{popularPages.edittext}}</a>
```
Comment (Grunny) [inline extensions/wikia/CommunityPage/templates/insightsPopularPages.mustache:10]: Whitespace issues.

File: extensions/wikia/CommunityPage/CommunityPageSpecialController.class.php [hunk 19]
```diff
@@ -43,10 +48,63 @@ public function index() {
...
+				'communitypage-noofviews',
+				$insight['metadata']['pv7']
+			)->text();
+			$insightsPages['pages'][$key]['editlink'] = $this->getEditUrl( $insight['link']['url'] );
+		}
+		return $insightsPages;
+	}
+
+	private function getEditUrl( $articleUrl ) {
+		if( EditorPreference::isVisualEditorPrimary() && $this->getUser()->isLoggedIn() ) {
```
Comment (Grunny) [inline extensions/wikia/CommunityPage/CommunityPageSpecialController.class.php]: Why are we only linking to VE if the user is logged in? Anonymous users should get VE as well, right?

Also, minor: there should be a space between `if` and the opening parenthesis.

**Comment by Grunny** [comment]
LGTM.

**Comment by adamkarminski** [comment]
Tested, LGTM
